### PR TITLE
feat: add BitVec.intMin

### PR DESCRIPTION
**Pull request by tobiasgrosser**
We keep the library interface for `intMin` minimal and make it an `abbrev` for `twoPow`. To ensure a uniform library interface, we also turn `intMax` into an abbrev for `allOnes`.

To improve proof automation, we introduce a simp-set `boolToPropSimps` plus corresponding lemmas that turn our `intMin_toLsb` proof into a simple `simp; omega` proof. This simp set is not yet active, but should become so after a stage-0 update.

This PR also pulls in some mathlib theorems on testBit and Nat and establishes facts about `2 ^ w` that are needed here and which are generally useful for bitvector reasoning.

The following theorem is not generalized to arbitrary `x` instead of `2`, as this would require a condition to be added for `x > 1` which would have to be passed to simp each time this theorem should fire:

```lean
protected theorem two_pow_pred_mod_two_pow (h : 0 < w) :
    2 ^ (w - 1) % 2 ^ w = 2 ^ (w - 1)
```

Mathlib reports only errors for theorems that this PR moves into lean.

```
error: ././././Mathlib/Data/Nat/Bitwise.lean:250:8: 'Nat.testBit_two_pow_self' has already been declared
error: ././././Mathlib/Data/Nat/Bitwise.lean:254:8: 'Nat.testBit_two_pow_of_ne' has already been declared
error: ././././Mathlib/Data/Nat/Bitwise.lean:265:8: 'Nat.testBit_two_pow' has already been declared
```

### CODE REVIEW

**Comment by tobiasgrosser** [comment]
Draft, as I am still polishing the order of the various theorems within the files. Any feedback is obviously already welcome.

File: src/Init/Data/BitVec/Lemmas.lean
Comment (hargoniX) [inline src/Init/Data/BitVec/Lemmas.lean:1458]: Non terminal simp

File: src/Init/Data/BitVec/Lemmas.lean
Comment (hargoniX) [inline src/Init/Data/BitVec/Lemmas.lean:1459]: There is a lemma for this kind of case distinction: [Nat.lt_trichotomy](https://leanprover-community.github.io/mathlib4_docs/Init/Data/Nat/Basic.html#Nat.lt_trichotomy)

File: src/Init/Data/BitVec/Lemmas.lean
Comment (AtticusKuhn) [inline src/Init/Data/BitVec/Lemmas.lean:1465]: could these two simps be combined into one simp?

**Review by AtticusKuhn — APPROVED** [review]
looks good to me.

File: src/Init/Data/BitVec/Lemmas.lean
Comment (AtticusKuhn) [inline src/Init/Data/BitVec/Lemmas.lean:1490]: How is "subst" different from "rw" or "simp"? (I tried to look up the documentation, did not find much). If "subst" just provides substitution, then maybe this proof could be shortened to 
```lean
simp [h, intMin]
```

Comment (hargoniX) [reply src/Init/Data/BitVec/Lemmas.lean:1489]: `subst` does indeed "just provide substitution". However in certain DTT hell scenarios `subst` is often able to perform substitutions along an equality while `simp` and `rw` cannot due to the way that they work. Whether that is the case here I do not know.

Comment (tobiasgrosser) [reply src/Init/Data/BitVec/Lemmas.lean:1489]: I tried, but here the order seems to be relevant.

**Comment by tobiasgrosser** [comment]
This is now ready-for-review from my side.

**Comment by tobiasgrosser** [comment]
awaiting-review

File: src/Init/Data/Nat/Bitwise/Lemmas.lean
Comment (semorrison) [inline src/Init/Data/Nat/Bitwise/Lemmas.lean:341]: Could you please derive these from a (non-`@[simp]`) `testBit_two_pow`, with an if statement for the RHS?

File: src/Init/Data/Nat/Lemmas.lean
Comment (semorrison) [inline src/Init/Data/Nat/Lemmas.lean:731]:
```suggestion
    2 ^ (w - 1) + 2 ^ (w - 1) = 2 ^ w := by
```

File: src/Init/Data/Nat/Lemmas.lean
Comment (semorrison) [inline src/Init/Data/Nat/Lemmas.lean:727]:
```suggestion
protected theorem pow_pred_lt_pow {x w : Nat} (h₁ : 1 < x) (h₂ : 0 < w) :
    x ^ (w - 1) < x ^ w :=
  Nat.pow_lt_pow_of_lt h₁ (by omega)
```

**Comment by semorrison** [comment]
As we are modifying `simp` lemmas for `Nat`, I would like to see Mathlib tested against this. If you rebase this PR onto `nightly-with-mathlib`, you should hopefully get a `builds-mathlib` or `breaks-mathlib` label within an hour.

File: src/Init/Data/BitVec/Lemmas.lean
Comment (semorrison) [inline src/Init/Data/BitVec/Lemmas.lean:1465]: I'm not super happy with this. I think `intMin` and `intMax` should be moved below `twoPow`.

Then use `def intMin (w : Nat) : BitVec w := twoPow w (w - 1)`, and `def intMax (w : Nat) : BitVec w := allOnes w`.

In fact both could be abbrevs, or just not get defined?

Every new def adds lots of missing API which is not keeping up.

Comment (semorrison) [reply src/Init/Data/BitVec/Lemmas.lean:1465]: Moreover, the proof of `getLsb_intMin` should be `by simp [intMin, boolToPropSimps]; omega`, for a simp set that we don't have yet that converts statements about `decide` into the corresponding propositional formulas.

File: src/Init/Data/Nat/Bitwise/Lemmas.lean
Comment (tobiasgrosser) [reply src/Init/Data/Nat/Bitwise/Lemmas.lean]: Done.

File: src/Init/Data/BitVec/Lemmas.lean
Comment (tobiasgrosser) [reply src/Init/Data/BitVec/Lemmas.lean]: Done.

**Comment by tobiasgrosser** [comment]
The comments are adressed, except for finding the correct imports for the new simp set. I will await advice from @bollu.

**Comment by tobiasgrosser** [comment]
The simp-set is also defined, so the remaining issue is a build failure in the nix build, which I am not sure how to debug.

File: src/Init/Data/Nat/Bitwise/Lemmas.lean
Comment (bollu) [inline src/Init/Data/Nat/Bitwise/Lemmas.lean:327]:
```suggestion
theorem testBit_two_pow {n m : Nat} : Nat.testBit (2 ^ n) m = if n = m then true else false := by
```
The `simp` normal form is `⊢ (2 ^ n).testBit m = decide (n = m)`, and I think this should perhaps even live in `Nat/Bitwise/Lemmas.lean`, but I leave that up to Kim :)

Comment (tobiasgrosser) [reply src/Init/Data/Nat/Bitwise/Lemmas.lean]: Resolved.

File: src/Init/Data/BitVec/Lemmas.lean
Comment (bollu) [inline src/Init/Data/BitVec/Lemmas.lean:1800]:
```suggestion
@[simp bv_toNat]
theorem toNat_intMax : (intMax w).toNat = 2 ^ w - 1 := by
```

File: src/Init/Data/BitVec/Lemmas.lean
Comment (bollu) [inline src/Init/Data/BitVec/Lemmas.lean:1797]:
```suggestion
@[simp]
theorem getLsb_intMax (w : Nat) : (intMax w).getLsb i = decide (i < w) := by
```

**Comment by tobiasgrosser** [comment]
It seems that the intMax definition was incorrect. I corrected it, but I haven't polished the proofs yet.

File: src/Init/Data/Nat/Lemmas.lean
Comment (tobiasgrosser) [inline src/Init/Data/Nat/Lemmas.lean:734]:
```suggestion
```

File: src/Init/Data/BitVec/Lemmas.lean
Comment (semorrison) [inline src/Init/Data/BitVec/Lemmas.lean:1826]:
```suggestion
@[simp, bv_toNat]
theorem toNat_intMax : (intMax w).toNat = 2 ^ (w - 1) - 1 := by
  simp only [intMax]
  by_cases h : w = 0
  · simp [h]
  · have h' : 0 < w := by omega
    rw [toNat_sub, toNat_twoPow, ← Nat.sub_add_comm (by simpa [h'] using Nat.one_le_two_pow),
      Nat.add_sub_assoc (by simpa [h'] using Nat.one_le_two_pow),
      Nat.two_pow_pred_mod_two_pow h', ofNat_eq_ofNat, toNat_ofNat, Nat.one_mod_two_pow h',
      Nat.add_mod_left, Nat.mod_eq_of_lt]
    have := Nat.two_pow_pred_lt_two_pow h'
    have := Nat.two_pow_pos w
    omega

@[simp]
theorem getLsb_intMax (w : Nat) : (intMax w).getLsb i = decide (i + 1 < w) := by
  rw [← testBit_toNat, toNat_intMax, Nat.testBit_two_pow_sub_one, decide_eq_decide]
  omega
```

File: src/Init/Data/Nat/Lemmas.lean
Comment (semorrison) [inline src/Init/Data/Nat/Lemmas.lean:731]:
```suggestion
protected theorem pow_pred_lt_pow {x w : Nat} (h₁ : 1 < x) (h₂ : 0 < w) :
    x ^ (w - 1) < x ^ w :=
  Nat.pow_lt_pow_of_lt h₁ (by omega)

protected theorem two_pow_pred_lt_two_pow {w : Nat} (h : 0 < w) :
    2 ^ (w - 1) < 2 ^ w :=
  Nat.pow_pred_lt_pow (by omega) h

```

**Comment by semorrison** [comment]
Remaining steps to activating `boolToPropSimps`:
* Move `BoolToPropSimps.lean` to `src/Init/Data/`.
* add `import Init.Data.BoolToPropSimps` to `src/Init/Data.lean`
* commit that
* `make -j32 -C build/release update-stage0`
* `git commit -m "chore: update stage0"`
  (it's essential the commit message is exactly that)
* now uncomment the `@[boolToPropSimps]` attributes, and hopefully you can use `simp only [boolToPropSimps]`.

(If you give me write access on your repo I'm happy to do these steps, too.)

**Comment by tobiasgrosser** [comment]
> Remaining steps to activating `boolToPropSimps`:
> 
> * Move `BoolToPropSimps.lean` to `src/Init/Data/`.
> * add `import Init.Data.BoolToPropSimps` to `src/Init/Data.lean`
> * commit that
> * `make -j32 -C build/release update-stage0`
> * `git commit -m "chore: update stage0"`
>   (it's essential the commit message is exactly that)
> * now uncomment the `@[boolToPropSimps]` attributes, and hopefully you can use `simp only [boolToPropSimps]`.
> 
> (If you give me write access on your repo I'm happy to do these steps, too.)

Cool. I gave you commit access. I guess its easier if you push this through.

**Pull request merged**
